logging: set 'log_backend_rtt:panic_mode' before 'log_backend_std_panic'

adfernandes opened the pull request:
Call the LOG_PANIC() macro/function in the default
assert_post_action handler, otherwise queued error
messages will not be displayed.

Signed-off-by: Andrew Fernandes <andrew@fernandes.org>

andrewboie commented:
I'm not sure I understand why this is necessary, can you provide some more detail?

LOG_PANIC() should already be invoked as a consequence of the subsequent `k_oops()` call.

adfernandes commented:
@andrewboie A failed `__ASSERT(...)` in application code will call `assert_post_action()`, which by default will call:
* `k_oops()` if `CONFIG_USERSPACE` is defined, but will call
* `k_panic()` if `CONFIG_USERSPACE` is **not** defined.

Now `k_oops()` **does** call `LOG_PANIC` in [`k_sys_fatal_error_handler`](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/fatal.c#L36-L41).

But `k_panic()` does **not** call `LOG_PANIC` in [`assert_post_action`](https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/os/assert.c#L36-L45).

So what was happening is my application was freezing (due to a programming error on my part) and it was a pain to track down because the error was logged, but not displayed as the kernel went into a hard-fault spin.

andrewboie commented:
wait a second...`k_panic()` and `k_oops()` have the same implementation, they both end up in `k_sys_fatal_error_handler()`. They both are macros to `z_except_reason()`.

I think the real problem is that you are seeing different behavior between `k_panic()` and `k_oops()`? That shouldn't be happening

> But k_panic() does not call LOG_PANIC in assert_post_action .

Nothing should be calling `LOG_PANIC()` in `assert_post_action()`. What is supposed to happen is that `assert_post_action()` calls `k_oops()` or `k_panic()` which, regardless of which, eventually ends up in `z_fatal_error()` and then shortly afterwards in `k_sys_fatal_error_handler()`.

So what seems to be broken here is that we are not eventually ending up in `k_sys_fatal_error_handler()?` what is happening when k_panic() is triggered?

adfernandes commented:
Aha! Thanks, @andrewboie - your comments made me double-check what I _thought_ was going on. 

So if the logger is in panic mode before either `k_oops()` or `k_panic()`, everything is logged as you would expect, the registers are dumped, and the system goes into a spin-and-halt loop right where it should, with `k_sys_fatal_error_handler()` getting called precisely once.

(I'm on a Nordic Cortex-M4 using the RTT logging backend, FYI.)

But if the logging backend is _not_ in a panic, then `k_sys_fatal_error_handler()` gets called **twice** due to a second assertion failure.

What I _think_ is happening is on the first `k_{oops|panic}`, we use the `svc 2` to handle the exception. But then the RTT backend does not think there is a panic yet because `panic_mode` is not set to `true` until _after_ all the flushing and so on.

But because of the `svc` call, we are essentially in an ISR and we trigger the `__ASSERT` in `z_sched_lock`. I need to verify all of this.

But basically we double `__ASSERT` in this case.

I'll try to verify this a little more cleanly tomorrow, but I **have** verified that `k_sys_fatal_error_handler` **does** get called twice after a failed `__ASSERT`.

adfernandes commented:
Alrighty. Let me recap everything I've found _and verified_ in this single comment, for clarity.

My configuration:
* Zephyr `v2.3.0-rc1-310-g265d2cfa69`
* ARM Cortex-M4 (Nordic nRF52832 custom board)
* `CONFIG_LOG=y`
* `CONFIG_LOG_PRINTK=y`
* `CONFIG_LOG_MINIMAL` is **not** defined
* `CONFIG_USERSPACE` is **not** defined
* `CONFIG_RTT_CONSOLE=y`
* `CONFIG_USE_SEGGER_RTT=y

Trigger:
* Fail an `__ASSERT(...)` in an application thread

Sequence of events, verified via debugger:
* `__ASSERT(false)` calls `assert_post_action`
* The default `assert_post_action` calls one of `k_(oops|panic)`
* `k_(oops|panic)` both trigger a software interrupt `SVC #2`
* **We are now in an "interrupt context"**
* Eventually we make it to `k_sys_fatal_error_handler`
* `k_sys_fatal_error_handler` calls `log_panic` since logging is **not** minimal
* `log_panic` calls `z_impl_log_panic`
* In `z_impl_log_panic`, `log_core.c::panic_mode` is false on entry
* `z_impl_log_panic` calls `log_backend_rtt.c::panic`
* `log_backend_rtt.c::panic_mode` is false at this point
* `log_backend_rtt.c::panic` calls `log_backend_std_panic`
* `log_backend_std_panic` calls `log_output_flush`
* Eventually RTT's `data_out_block_mode` is called
* Now `data_out_block_mode` checks to see if the RTT backend [`!is_sync_mode()`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_backend_rtt.c#L207)
* But `!is_sync_mode()` is equivalent to `log_core.c::panic_mode` (as per the assembly)
* **Note** that `log_core.c::panic_mode` is `false` since we are _setting up_ the first panic
* Therefore `!is_sync_mode()` is `true` and the RTT backend calls `RTT_LOCK()`
* `RTT_LOCK` calls `z_impl_k_mutex_lock`
* The RTT backend calls `SEGGER_RTT_WriteSkipNoLock`
* Then we call `z_impl_k_mutex_unlock`
* `z_impl_k_mutex_unlock` passes [`__ASSERT_NO_MSG(mutex->lock_count > 0U)`](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/mutex.c#L230)
* `z_impl_k_mutex_unlock` calls [`z_sched_lock`](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/mutex.c#L233)
* `z_sched_lock` **fails** the [`!arch_is_in_isr()`](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/include/ksched.h#L267)
* ... and we call `assert_post_action` again, erroneously looping

The reason this PR makes the problem go away is because it sets the logger into panic mode early, so `RTT_(LOCK|UNLOCK)` are **not** called from an interrupt context.

Of course, that's not the correct solution to this logic bug. It does get rid of it, but... **is that the correct thing to do?** (Especially since `assert_post_action` is supposed to be application-replaceable, and may choose to not call `log_panic`...) 🤷‍♂️

adfernandes committed:
logging: set 'log_backend_rtt:panic_mode' before 'log_backend_std_panic'

Set 'log_backend_rtt:panic_mode' before calling 'log_backend_std_panic',
because otherwise the RTT backend behaves as if the system is NOT
panicking and tries to do asynchronous writes via RTT_LOCK/UNLOCK.

Signed-Off-By: Andrew Fernandes <andrew@fernandes.org>

adfernandes commented:
Okay. I've waded through the logic of this and I _think_ the bug lies in `log_backend_rtt.c`.

There is a `log_core.c::panic_mode` variable that is set to `true` **after** all the backends have been set into panic mode.

But  `log_backend_rtt.c` **also** has a `log_backend_rtt.c::panic_mode` variable, and I think that it needs to be set **before** the backend is fully notified of the panic.

Basically, I think we need to [swap the order of these two statements](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_backend_rtt.c#L260-L261).

I've updated this PR to do this, and this resolves my issue.

Please let me know what you think.

andrewboie commented:
@adfernandes thanks. I believe @nordic-krch is the expert in this area, I'll defer to his opinion.

nordic-krch reviewed (approved):
@adfernandes thanks for analysis. You are right about order swapping.

nordic-krch commented:
@carlescufi would be good to get that one in. Do we need a bug issue?

carlescufi commented:
Thanks for your persistence and analysis @adfernandes!
